libraries/AC_Avoidance/AP_OABendyRuler.cpp

```diff
@@ -0,0 +1,294 @@
+                // replace bearing with one that is closer to our current ground course
+                best_bearing = bearing_test;
+            }
+
+            // perform second stage test in three directions looking for obstacles
+            const float test_bearings[] = { 0, 45, -45 };
+            const float bearing_to_dest2 = test_loc.get_bearing_to(destination) * 0.01f;
+            float distance2 = constrain_float(lookahead_step2_dist, OA_BENDYRULER_LOOKAHEAD_STEP2_MIN, test_loc.get_distance(destination));
+            for (uint8_t j = 0; j < ARRAY_SIZE(test_bearings); j++) {
+                float bearing_test2 = wrap_180(bearing_to_dest2 + test_bearings[j]);
+                Location test_loc2 = test_loc;
+                test_loc2.offset_bearing(bearing_test2, distance2);
+
+                // calculate minimum margin to fence and obstacles for this scenario
+                float margin2 = calc_avoidance_margin(test_loc, test_loc2);
+                if (margin2 > _margin_max) {
+                    // all good, now project in the chosen direction by the full distance
+                    destination_new = current_loc;
+                    destination_new.offset_bearing(bearing_test, distance_to_dest);
+                    _current_lookahead = MIN(_lookahead, _current_lookahead * 1.1f);
+                    return (i != 0 || j != 0);
+                }
+            }
+        }
+    }
+
+    float chosen_bearing;
+    if (have_best_bearing) {
+        // none of the directions tested were OK for 2-step checks. Choose the direction
+        // that was best for the first step
+        chosen_bearing = best_bearing;
+        _current_lookahead = MIN(_lookahead, _current_lookahead * 1.05f);
+    } else {
+        // none of the possible paths had a positive margin. Choose
+        // the one with the highest margin
+        chosen_bearing = best_margin_bearing;
+        _current_lookahead = MAX(_lookahead * 0.5f, _current_lookahead * 0.9f);
+    }
+
+    // calculate new target based on best effort
+    destination_new = current_loc;
+    destination_new.offset_bearing(chosen_bearing, distance_to_dest);
+
+    return true;
+}
+
+// calculate minimum distance between a path and any obstacle
+float AP_OABendyRuler::calc_avoidance_margin(const Location &start, const Location &end)
+{
+    float circular_fence_margin;
+    if (!calc_margin_from_circular_fence(start, end, circular_fence_margin)) {
+        circular_fence_margin = FLT_MAX;
+    }
+
+    float polygon_fence_margin;
+    if (!calc_margin_from_polygon_fence(start, end, polygon_fence_margin)) {
+        polygon_fence_margin = FLT_MAX;
+    }
+
+    float proximity_margin;
+    if (!calc_margin_from_proximity_sensors(start, end, proximity_margin)) {
+        proximity_margin = FLT_MAX;
+    }
+
+    // return smallest margin from any obstacle
+    return MIN(MIN(circular_fence_margin, polygon_fence_margin), proximity_margin);
```

Comment:
This is going to get nasty :-)

Next time somethings added we probably want a `float smallest_margin = FLT_MAX` at the top of this function.  Each clause in here would stop negating the result of the method call and instead do the `MIN(...)` inside the body of the `if`.

Reply:
ok, I guess we could change the functions so they don't return a boolean and instead return FLT_MAX if they can't calculate a margin for some reason (i.e. the fence doesn't exist).